### PR TITLE
Implement Vim like movement commands

### PR DESCRIPTION
**Pull request by mitsuhiko**
This adds a few new movement commands to allow `hjkl` to be rebound so that they follow more the movement of vim. left/right movement cannot move between lines and up/down either follows newline or does not accidentally move onto the newline character. See #4268 for motivation. This however does not emulate vim, it just has some vim inspired movements. In particular it retains the entire behavior of Helix with regards to how newlines are otherwise handled (they stay selectable, and these alternative movement commands to not interfere with other commands in use or change their assumptions).

vim-movement mov

The following commands are added:

* `move_line_up_anchored` (bind to `k`): implements anchored movement up
* `move_line_down_anchored` (bind to `j`): implements anchored movement down
* `move_char_left_same_line` (bind to `h`): implements same-line movement left
* `move_char_right_same_line` (bind to `l`): implements same-line movement right
* `extend_line_up_anchored` (bind to `k`): implements anchored selection extension up
* `extend_line_down_anchored` (bind to `j`): implements anchored selection extension down
* `extend_char_left_same_line` (bind to `h`): implements same-line selection extension left
* `extend_char_right_same_line` (bind to `l`): implements same-line selection extension right

## Bindings

```toml
[keys.normal]
j = ["move_line_down_anchored"]
k = ["move_line_up_anchored"]
h = ["move_char_left_same_line"]
l = ["move_char_right_same_line"]

[keys.select]
j = ["extend_line_down_anchored"]
k = ["extend_line_up_anchored"]
h = ["extend_char_left_same_line"]
l = ["extend_char_right_same_line"]
```

## Same Line Movement

The default movement behavior of helix will happily place you in the next line. Vim users are not used to that and are likely to prefer the cursor not to move into the next line (that is at least my expectation). This is what `move_char_left_same_line` and `move_char_right_same_line` do. They will however let you move to the rightmost character (the newline) which is not what vim does (except of `virtualedit=oneline` is used). However this pairs up nicely with the anchored movement (see next).

## Anchored Vertical Movements

The two commands `move_line_up_anchored` and `move_line_down_anchored` implement what I would call anchored movement. If you are on the newline character (eg: one to the right of where `goto_line_end` would place you) then as you move up and down you end up on the newline of the target line again.  This emulates the behavior of vim of what happens if you move to the line end with `$`. If however you are placed to the left of the newline, as you move up and down you will not move onto the newline.

The main difference is that in vim if you use `$` and go up/down you stay anchored to the line end. In Helix with this change you need to do `gll` to stay anchored to the newline. I do however have to say that I prefer this actually because in Vim you do not actually know if you are newline anchored visually. That is as far as I can tell an internal state.

I played around with this a bit now and I really like the combination of the two. It stays true to the helix character model, but it also feels very natural for vim users.

**Limitation:** Movements across empty lines are ambiguous. To help with this I put the horizontal column to the largest possible integer on the range to tell these cases apart. From what I can tell this does not create negative impact and avoids extra complication on the range.

## Notes on Mixed Use

I think these move/extend commands are interesting to use independent of each other. For instance I find the anchored newline movement functionality useful even if one does not like restricting the movements within a line. For instance line based selection works really nice as once the visual selection is on the newline, any further moving down will continue to capture entire lines. That means for instance `vx` will continue to move downwards which I think also addresses some points of #1638 as it basically keeps you in line mode for as long as you don't move left.

### CODE REVIEW

**Comment by the-mikedavis** [comment]
Newlines being selectable like any other character is intended behavior and I don't like adding options or alternate movement commands to skirt them. See https://github.com/helix-editor/helix/issues/2956 and https://github.com/helix-editor/helix/issues/3076.

**Comment by mitsuhiko** [comment]
@the-mikedavis this does not change the behavior about select-ability. They are still selectable as before. The new commands add a way that for horizontal movements you can stay constrained within the line but you still move onto them and for vertical movement it even anchors you onto them if you already are on one.

**Comment by archseer** [comment]
(Also fixes https://github.com/helix-editor/helix/issues/768)

**Comment by mitsuhiko** [comment]
Curious about what the odds are of this merging. I'm happy to incorporate necessary changes if there is something I can do here.

File: helix-core/src/movement.rs [hunk 1]
```diff
@@ -80,6 +102,84 @@ pub fn move_vertically(
...
+    // Compute the current position's 2d coordinates.
+    let Position { row, col } = visual_coords_at_pos(slice, pos, tab_width);
+    let horiz = range.horiz.unwrap_or(col as u32);
+
+    // if we are resting on a newline character and we did not just move
+    // across an empty line we are in newline movement mode.  To disambiugate these
+    // cases we take advantage that anchored movement sets the horizontal column
+    // to the largest possible integer.
+    let newline_move_mode = pos == line_end_char_index(&slice, line)
+        && (horiz == !0 || !rope_is_line_ending(slice.line(line)));
```
Comment (the-mikedavis) [inline helix-core/src/movement.rs:134]:
```suggestion
        && (horiz == u32::MAX || !rope_is_line_ending(slice.line(line)));
```

File: helix-core/src/movement.rs [hunk 2]
```diff
@@ -80,6 +102,84 @@ pub fn move_vertically(
...
+            slice.char_to_line(pos_at_visual_coords(
+                slice,
+                Position::new(new_row, col),
+                tab_width,
+            )),
+        );
+
+        // when in newline_move_mode we set the horizontal column to the largest
+        // possible integer to be able to disambiugate movements across empty lines.
+        (new_pos, !0)
```
Comment (the-mikedavis) [inline helix-core/src/movement.rs:156]:
```suggestion
        (new_pos, u32::MAX)
```

**Review by the-mikedavis — COMMENTED** [review]
Looking at your comment here (https://github.com/helix-editor/helix/pull/4204#issuecomment-1279345053), Vim whichwrap and the context in https://github.com/helix-editor/helix/issues/768, I think these motions are ok and don't introduce incompatible Vim-isms 👍. (Sorry for jumping the gun on "no Vim" here; steam preemptively comes out of my ears when I see "Add Vim-like..." 😅)

**Comment by mitsuhiko** [comment]
An open question I have is how this should be documented best. Is there a place to document unmapped commands that do not have a typable alias?

**Comment by the-mikedavis** [comment]
Not currently, see: https://github.com/helix-editor/helix/pull/997

**Comment by mitsuhiko** [comment]
Then I suppose this is ready.

**Review by the-mikedavis — APPROVED** [review]

**Comment by archseer** [comment]
Sorry for the delay, this looks fine to me but needs a rebase now that the soft wrap changes landed!

**Comment by pascalkuthe** [comment]
The rebase will be slightly more challenging than usual because the vertical movement commands were essentially reimplemented in #5420. It would also be good to have a version that does visual line movement and document line movement like we have for the normal movement command

**Comment by mitsuhiko** [comment]
I will update this. Unfortunately the changes are a bit more involved so it will take a bit until I find the time for it.

**Comment by yashpalgoyal1304** [comment]
i was also thinking about this _topic_ the other day - that why don't n/vi/m stop the horizontal movement at line ends and don't wrap it around - is there any benefit to that way while editing etc...

**Comment by mkblast** [comment]
I think a better way to configure it is to add all of those binding into one option.

File: helix-core/src/movement.rs [hunk 3]
```diff
@@ -80,6 +102,84 @@ pub fn move_vertically(
...
+/// Implements anchored vertical movement.
+///
+/// Anchored movement behaves differently depending on where the cursor is placed with
+/// regards to newlines.  If the cursor is currently on a newline, then moving up and
+/// down will place you again on the newline of that line.
+///
+/// If you are not on the newline, then moving up and down will move you to the same
+/// column except if that column is further to the left from where you are.  In that case
+/// it places you on the rightmost column that is not a newline.
+pub fn move_vertically_anchored(
```
Comment (pickfire) [inline helix-core/src/movement.rs:114]: Can we have some tests for both of these functions?

**Comment by pascalkuthe** [comment]
I would definitely like to still see this feature. Unfortunately, this branch is quite stale as the refactor in #5420 entirely changed how movement works so somebody would need to start almost from scratch. Thank you for contributing!

**Pull request closed without merging**